### PR TITLE
#29662:  Introduce tor_assertf{_nonfatal}() to allow logging extra error message on assert failure

### PR DESCRIPTION
**Pull request by rl1987**
https://trac.torproject.org/projects/tor/ticket/29662

### CODE REVIEW

File: src/lib/log/util_bug.c [hunk 1]
```diff
@@ -69,23 +69,43 @@ tor_set_failed_assertion_callback(void (*fn)(void))
 
 /** Helper for tor_assert: report the assertion failure. */
 void
+__attribute__((format(printf, 5, 6)))
 tor_assertion_failed_(const char *fname, unsigned int line,
-                      const char *func, const char *expr)
+                      const char *func, const char *expr,
+                      const char *fmt, ...)
 {
   char buf[256];
```
Comment (teor2345) [inline src/lib/log/util_bug.c]: Is buf large enough for arbitrary printf strings and arguments?

Comment (teor2345) [reply src/lib/log/util_bug.c]: We could use tor_asprintf() instead?

File: changes/ticket29662 [hunk 2]
```diff
@@ -0,0 +1,5 @@
+  o Minor features (debugging):
+    - Introduce tor_assertf() and tor_assertf_nonfatal() to enable logging of
+      additional information during assert failure. Now we can use format
+      strings to include pieces of information that is relevant for trouble
```
Comment (teor2345) [inline changes/ticket29662]: Grammar nitpick:
```suggestion
      strings to include pieces of information that are relevant for trouble
```

File: src/lib/log/util_bug.c [hunk 3]
```diff
@@ -69,23 +69,43 @@ tor_set_failed_assertion_callback(void (*fn)(void))
...
 
 /** Helper for tor_assert_nonfatal: report the assertion failure. */
 void
+__attribute__((format(printf, 6, 7)))
 tor_bug_occurred_(const char *fname, unsigned int line,
                   const char *func, const char *expr,
-                  int once)
+                  int once, const char *fmt, ...)
 {
   char buf[256];
```
Comment (teor2345) [inline src/lib/log/util_bug.c]: Is buf large enough?
Should we use tor_asprintf?

File: src/lib/log/util_bug.c (hunk 1 above)
Comment (rl1987) [reply src/lib/log/util_bug.c]: Doing so in b53a0bc2d1432acb2c8d028461ca6eae2a9b93bc.

File: src/lib/log/util_bug.c (hunk 3 above)
Comment (rl1987) [reply src/lib/log/util_bug.c]: b53a0bc2d1432acb2c8d028461ca6eae2a9b93bc

File: changes/ticket29662 (hunk 2 above)
Comment (rl1987) [reply changes/ticket29662]: Fixed in 34c6eb25d88ad214ca5412b04573a3a1f5404de8.

**Pull request closed without merging**